src/utils/IdealGraphVisualizer/HierarchicalLayout/src/main/java/com/sun/hotspot/igv/hierarchicallayout/HierarchicalStableLayoutManager.java

```diff
@@ -0,0 +1,1909 @@
+                                    }
+                                    int compToXCoord = e.to.x;
+                                    if (e.to.vertex != null) {
+                                        compToXCoord += e.relativeTo;
+                                    }
+                                    if ((fromNodeXCoord > compFromXCoord && toNodeXCoord < compToXCoord)
+                                            || (fromNodeXCoord < compFromXCoord
+                                            && toNodeXCoord > compToXCoord)) {
+                                        currentCrossings += 1;
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+                if (currentCrossings <= edgeCrossings) {
+                    edgeCrossings = currentCrossings;
+                    optimalPos = i;
+                }
+            }
+            assert optimalPos != -1;
+            return optimalPos;
+        }
+
+        /**
+         * Insert node at the assigned layer, updating the positions of the nodes within
+         * the layer
+         *
+         * @param node
+         * @param layer
+         */
+        private void insertNode(LayoutNode node, int layer) {
+            assert layers.keySet().contains(layer) || layer == 0;
+
+            node.layer = layer;
+            List<LayoutNode> layerNodes = layers.getOrDefault(layer, new ArrayList<LayoutNode>());
+
+            if (layerNodes.size() == 0) {
+                node.pos = 0;
+            } else {
+                node.pos = optimalPosition(node, layer);
+            }
+
+            for (LayoutNode n : layerNodes) {
+                if (n.pos >= node.pos) {
+                    n.pos += 1;
+                }
+            }
+            layerNodes.add(node);
+            layers.put(layer, layerNodes);
+
+            if (!nodes.contains(node)) {
+                nodes.add(node);
+            }
+            if (node.vertex != null) {
+                vertexToLayoutNode.put(node.vertex, node);
+            }
+
+            adjustXCoordinates(layer);
+
+            sanityCheckNodesAndLayerNodes();
+        }
+
+        private void processSingleEdge(LayoutEdge e) {
+            LayoutNode n = e.to;
+            if (e.to.layer - 1 > e.from.layer) {
+                LayoutEdge last = e;
+                for (int i = n.layer - 1; i > last.from.layer; i--) {
+                    last = addBetween(last, i);
+                }
+            }
+        }
+
+        private LayoutEdge addBetween(LayoutEdge e, int layer) {
+            LayoutNode n = new LayoutNode();
+            n.width = DUMMY_WIDTH;
+            n.height = DUMMY_HEIGHT;
+            n.succs.add(e);
+            LayoutEdge result = new LayoutEdge();
+            result.vip = e.vip;
+            n.preds.add(result);
+            result.to = n;
+            result.relativeTo = n.width / 2;
```

Comment:
Yes since `relativeTo` and `relativeFrom` refers to the ports on the node, which is irrelevant for the dummy nodes. Could definitely be done differently, but this is how it is in `HierarchicalLayoutManager` and I thought it would be better to be consistent across the layout managers

Reply:
Okay, thanks for the explanation. The code in `HierarchicalLayoutManager` looks very similar. So, you could also change `relativeTo` and `relativeFrom` to zero there. Or even better share the code somehow (if I see that correctly, the only difference is how to insert the node - `nodes.add(n)` vs. `insertNode(n, layer)`).

Reply:
@robcasloz has created a suggestion list of how to improve `HierarchicalLayoutManager`. Could we perhaps add this to that list instead? I think there must have been some reason to set `n.width/2` (maybe just in case you want to make the dummy nodes visible with a larger width) so it could be worth to work it through some more

Reply:
I'm fine with moving this and my other suggested clean-ups to this list. However, even after removing `sanityCheckNodesAndLayerNodes()` and additionally removing the `assert` checks (which are currently still in the code), the example graph mentioned earlier still needs around 18s to load. The bottleneck is now `sanityCheckEdges()`.